liblepton/scheme/lepton/object.scm

```diff
@@ -493,45 +530,171 @@ returns #f."
+                     (bytevector-sint-ref ymin 0 (native-endianness) (sizeof int)))))))
+
 
 (define-public (fold-bounds . bounds)
   (fold
    (lambda (a b)
      (if (and a b)
-         ;; calc bounds
+         ;; Calculate bounds.
          (cons (cons (min (caar a) (caar b))    ; left
                      (max (cdar a) (cdar b)))   ; top
                (cons (max (cadr a) (cadr b))    ; right
-                     (min (cddr a) (cddr b)))) ; bottom
+                     (min (cddr a) (cddr b))))  ; bottom
 
-         ;; return whichever isn't #f
+         ;; Return whichever isn't #f.
          (or a b)))
-   #f ;; default
+   ;; Default.
+   #f
    bounds))
 
+(define (object-bounds . objects)
+  "Returns the bounds of the objects in the variable-length
+argument list OBJECTS.  The bounds are returned as a pair
+structure of the form: ((left . top) . (right . bottom)).  If the
+list OBJECTS is empty, or none of the objects has any
+bounds (e.g. because they are all empty components and/or text
+strings), returns #f.  Warning: This function always returns the
+actual bounds of the objects, not the visible bounds."
+  (apply fold-bounds (map one-object-bounds objects)))
+
+
 ;;;; Object transformations
 
 (define-public (translate-objects! vector . objects)
   (for-each
-   (lambda (x) (%translate-object! x (car vector) (cdr vector)))
+   (lambda (x) (translate-object! x (car vector) (cdr vector)))
    objects)
   objects)
 
 (define-public (rotate-objects! center angle . objects)
   (for-each
-   (lambda (x) (%rotate-object! x (car center) (cdr center) angle))
+   (lambda (x) (rotate-object! x (car center) (cdr center) angle))
    objects)
   objects)
 
 (define-public (mirror-objects! x . objects)
   (for-each
-   (lambda (obj) (%mirror-object! obj x))
+   (lambda (obj) (mirror-object! obj x))
    objects)
   objects)
 
-( define-public object-selectable?     %object-selectable? )
-( define-public set-object-selectable! %set-object-selectable! )
 
-( define-public object-embedded?     %object-embedded? )
-( define-public set-object-embedded! %set-object-embedded! )
+(define (object-selectable? object)
+  "Checks the state of OBJECT's selectable flag: if it's true, the
+object is considered to be unlocked, otherwise it is locked.
+Returns #t if OBJECT is selectable, otherwise returns #f."
+  (define pointer (geda-object->pointer* object 1))
+
+  (true? (lepton_object_get_selectable pointer)))
+
+
+(define (set-object-selectable! object selectable?)
+  "Sets OBJECT's selectable flag to SELECTABLE?, thus locking or
+unlocking OBJECT.  Locked objects cannot be selected in GUI.
+Returns OBJECT."
+  (define pointer (geda-object->pointer* object 1))
+
+  (unless (eq? (object-selectable? object) selectable?)
+    (lepton_object_set_selectable pointer (if selectable? 1 0))
+    (lepton_object_page_set_changed pointer))
+
+  object)
+
+
+
+(define (object-embedded? object)
```

Comment:
Looking at the code of the brand-new `object-embedded?()` function,
the only conclusion we can make is this: _Пить надо меньше_. :)))

Reply:
Please don't get me wrong, the point is that experimenting
with new fancy-pancy Guile features could be done in another
repository (e.g. your ingenious (and I mean it) gschem-goodies).

Reply:
> the only conclusion we can make is this: _Пить надо меньше_. :)))

...сказал Дима, добивая 0,7 :-))

Reply:
> Looking at the code of the brand-new `object-embedded?()` function,
> the only conclusion we can make is this: _Пить надо меньше_. :)))

What's wrong with the function?

Reply:
> Please don't get me wrong, the point is that experimenting
> with new fancy-pancy Guile features could be done in another
> repository (e.g. your ingenious (and I mean it) gschem-goodies).

The point is to get rid of so named `core` modules defined in C.  This will let us reduce and simplify code of the tools and enable *native* compilation using `guild`, I believe.  It's not just experimenting with Guile, it's the direction I've intentionally chosen.  Please be more constructive in your criticism.

Reply:
> > the only conclusion we can make is this: _Пить надо меньше_. :)))
> 
> ...сказал Дима, добивая 0,7 :-))

Встану рано утром, посмотрю на рожу,
Больше пить не буду... Но и меньше тоже.

Reply:
> Встану рано утром, посмотрю на рожу,
> Больше пить не буду... Но и меньше тоже.

Brilliant!

Reply:
Вот ты язва, Володя!
Судя по всему - служил в ВДВ.
Вываливай уже фотку!

Reply:
Я не служил, сдал кандидатский минимум, пошёл в аспирантуру, потом бросил, детей нарожал.

Reply:
И ты прав, так держать!
Это ведь у меня ни кола ни двора.
Только водка и осталась.